manual/src/tutorials/effectexamples.etex

```diff
@@ -0,0 +1,666 @@
+effect, we enqueue the current task and run the next one from the scheduler
+queue. If the effect is "Fork f", we enqueue the current task, and recursively
+call "spawn" on "f", which run the given task in its own handler effectively
+running the task concurrently.
+
+If the effect is "Xchg", then we first check whether there is a task waiting to
+exchange. If so, we enqueue the waiting task with the current value being
+offered, and immediately resume the current task with the value being offered.
+If not, we make the current task the waiting exchanger, and run the next task
+from the scheduler queue.
+
+Now we can write a concurrent program that utilises the newly defined
+operations:
+
+\begin{caml_example}{verbatim}
+open Printf
+
+let _ = run (fun _ ->
+  fork (fun _ ->
+    printf "[t1] Sending 0\n";
+    printf "[t1] received %d\n" (xchg 0));
+  fork (fun _ ->
+    printf "[t2] Sending 1\n";
+    printf "[t2] received %d\n" (xchg 1)))
+\end{caml_example}
+
+Observe that the messages from the two tasks are interleaved. Notice also that
+the snippet above makes no reference to the effect handlers, and is in
+direct-style (no monadic operations). This example illustrates that, with
+effect handlers, the user code in a concurrent program can remain in simple
+direct-style, and the use of effect handlers can be fully contained within the
+concurrency library implementation.
+
+\subsection{s:effects-discontinue}{Resuming with an exception}
+
+In addition to resuming a continuation with a value, effect handlers also
+permit resuming by raising an effect at the point of perform. This is done with
+the help of "discontinue" primitive. Discontinue helps ensure that
+resource-safe code in OCaml behaves correctly.
+
+For example, consider the dequeue operation in the previous example reproduced
+below:
+
+\begin{caml}
+\begin{camlinput}
+let dequeue () =
+  if Queue.is_empty run_q then () (* done *)
+  else (Queue.pop run_q) ()
+\end{camlinput}
+\end{caml}
+
+If the scheduler queue is empty, dequeue considers the scheduler to be complete
+and returns to the caller. However, there may still be a pending task waiting
+to exchange a value, which remains blocked forever!. If the blocked task holds
+onto resources, such resources are leaked. For example, consider the following
+task:
+
+\begin{caml_example*}{verbatim}
+let leaky_task () =
+  fork (fun _ ->
+    let oc = open_out "secret.txt" in
+    Fun.protect ~finally:(fun _ -> close_out oc) (fun _ ->
+      output_value oc (xchg 0)))
+\end{caml_example*}
+
+The task writes the received message to the file "secret.txt". It uses
+"Fun.protect" to ensure that the output channel "oc" is closed on both normal
+and exceptional return cases. If this task does not have a matching exchange,
+then the task will remain blocked forever, and leaks the output channel "oc".
+
+We can use "discontinue" to ensure that the blocked task gets a chance to free
```

Comment:
I would suggest changing this sentence; "gets a chance" suggests that we are not 100% sure that this will work. Suggestion: "To avoid this problem, one must adhere to a simple discipline: **every continuation must be eventually either continued or discontinued**. Here, we use `discontinue` to ensure that the blocked task does not remain blocked forever. By discontinuing this task, we force it to terminate (with an exception)."